### PR TITLE
Add K-Nearest Neighbors

### PR DESCRIPTION
**Pull request by msluszniak**
I need to add tests

### CODE REVIEW

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    ],
+    metric: [
+      type: {:in, [:minkowski, :cosine]},
+      default: :minkowski,
+      doc: ~S"""
+      Name of the metric. Possible values:
+      * `:minkowski` - Minkowski metric. By changing value of `:p` parameter we can set Euclidean, Chebyshev, Manhattan or any arbitrary $L_p$ metric.
+
+      * `:cosine` - Cosine metric.
+      """
```
Comment (josevalim) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:52]: What if we merge `metric` and `p`? We could do `metric: :cosine` and `metric: {:minkowski, 1}`?

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+
+    * `:default_num_neighbors` - The number of neighbors to use by default for `k_neighbors` queries.
+
+    * `:weights` - Weight function used in prediction.
+
+    * `:num_classes` - Number of classes in provided labels.
+
+  ## Examples
+
+      iex>  Scholar.Neighbors.KNearestNeighbors.fit(Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]]), Nx.tensor([1, 0, 1, 1]),
```
Comment (josevalim) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:98]:
```suggestion
      iex> foo = Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]])
      iex>  Scholar.Neighbors.KNearestNeighbors.fit(foo, Nx.tensor([1, 0, 1, 1]),
```

But please give it a better name to the tensor. :D

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+  Currently 2D labels only supported for regression task.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
```
Comment (josevalim) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:84]:
```suggestion
  The function returns a struct with the following parameters:
```

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+  Find the K-neighbors of a point.
+
+  ## Return Values
+
+    Returns indices of and distances to the neighbors of each point.
+
+  ## Examples
+
+      iex> model =
+      iex>  Scholar.Neighbors.KNearestNeighbors.fit(Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]]), Nx.tensor([1, 0, 1, 1]),
```
Comment (josevalim) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:273]: Same here, let's define the tensor separately so the code is less overloaded. :)

**Review by josevalim — APPROVED** [review]
LGTM! I have only added some minor notes.

Btw, is this a good candidate for a notebook? :)

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+      end
+
+    indices =
+      Nx.stack(
+        [Nx.iota(Nx.shape(pred_labels), axis: 0), Nx.take(labels, pred_labels)],
+        axis: -1
+      )
+      |> Nx.flatten(axes: [0, 1])
+
+    proba = Nx.indexed_add(proba, indices, Nx.flatten(weights_vals))
```
Comment (polvalente) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:259]:
```suggestion
    proba = Nx.indexed_put(proba, indices, Nx.flatten(weights_vals))
```

indexed_put seems a bit more semantic given it's all zeros anyway

Comment (polvalente) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:259]: nevermind this comment :)

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+    {tensor, weights, axis} =
+      if axis == 0,
+        do: {Nx.transpose(tensor), Nx.transpose(weights), 1},
+        else: {tensor, weights, axis}
+
+    {num_samples, num_features} = tensor_shape = Nx.shape(tensor)
+
+    indices = Nx.argsort(tensor, axis: axis)
+
+    sorted = Nx.sort(tensor, axis: axis)
```
Comment (polvalente) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:351]:
```suggestion
    indices = Nx.argsort(tensor, axis: axis)

    sorted = Nx.take_along_axis(tensor, indices, axis: axis)
```

Comment (polvalente) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:351]: This avoids sorting twice

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+    counting_indices =
+      [
+        Nx.shape(group_indices)
+        |> Nx.iota(axis: 0)
+        |> Nx.reshape({num_elements, 1}),
+        Nx.reshape(group_indices, {num_elements, 1})
+      ]
+      |> Nx.concatenate(axis: 1)
+
+    to_add = Nx.broadcast(1, {num_elements}) * (weights |> Nx.reshape({:auto}))
```
Comment (polvalente) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:379]: Isn't this the same as `Nx.flatten(weights)`?

I mean, if the shapes differ anyway, the multiplication won't work as there is no `1`-sized axis in neither tensor for broadcasting, so in the end this is just an element-wise multiplication by 1

Comment (msluszniak) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:379]: Ofc you're right, I will fix that

**Comment by msluszniak** [comment]
> LGTM! I have only added some minor notes.
> 
> Btw, is this a good candidate for a notebook? :)

I think I can try to create a notebook with KNN as a major theme. :)

File: lib/scholar/neighbors/k_nearest_neighbors.ex (hunk 1 above)
Comment (msluszniak) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:52]: Yes, we could do that. The only thing that convinces me to stay with `:metric` and `:p` separately is that the check with NimbleOptions is simpler for this setup.

Comment (josevalim) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:52]: That's a good point but I think my suggestion has a better user experience and documentation, so please let's go with that!

File: lib/scholar/options.ex [hunk 8]
```diff
@@ -41,4 +41,16 @@ defmodule Scholar.Options do
...
+    if metric == :cosine or
+         (is_tuple(metric) and tuple_size(metric) == 2 and elem(metric, 0) == :minkowski and
+            is_number(elem(metric, 1)) and
+            elem(metric, 1) >= 0) do
+      {:ok, metric}
+    else
+      {:error,
+       "expected metric to be a :cosine or tuple {:minkowski, p} where p is non-negative number, got: #{inspect(metric)}"}
+    end
+  end
```
Comment (josevalim) [inline lib/scholar/options.ex:55]:
```suggestion
  def metric(:cosine), do: :ok
  def metric({:minkowski, p}) when is_integer(p) and p >= 0, do: :ok

  def metric(metric) do
    {:error,
     "expected metric to be a :cosine or tuple {:minkowski, p} where p is non-negative number, got: #{inspect(metric)}"}
  end
```

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+        * `:distance` - weight points by the inverse of their distance. in this case, closer neighbors of
+          a query point will have a greater influence than neighbors which are further away.
+      """
+    ],
+    metric: [
+      type: {:custom, Scholar.Options, :metric, []},
+      default: {:minkowski, 2},
+      doc: ~S"""
+      Name of the metric. Possible values:
+      * `{:minkowski, p}` - Minkowski metric. By changing value of `p` parameter we can set Euclidean, Chebyshev, Manhattan or any arbitrary $L_p$ metric.
```
Comment (josevalim) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:41]:
```suggestion
      Name of the metric. Possible values:

      * `{:minkowski, p}` - Minkowski metric. By changing value of `p` parameter (an integer)
        we can set Manhattan (1), Euclidean (2), Chebyshev, or any arbitrary $L_p$ metric.
```

However, isn't  "Chebyshev" infinity? How would we set it in this case?

**Review by josevalim — COMMENTED** [review]
Two minor nits and we can ship it!

File: lib/scholar/neighbors/k_nearest_neighbors.ex (hunk 9 above)
Comment (msluszniak) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:41]: Yeah, we should be able to pass `:inf` as `:p`. In the `Scholar.Metrics.Distance` there is currently an error that computes Chebyshev for p = 0 which is wrong. I will fix this and then change it here as well.

Comment (msluszniak) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:41]: And p is actually not an integer in general, just a positive float. In sklearn it is indeed an integer, but I really don't know why.

Comment (josevalim) [reply lib/scholar/neighbors/k_nearest_neighbors.ex:41]: Ah, got it. So let’s make it a float in the validations and support `:infinity`.

File: lib/scholar/neighbors/k_nearest_neighbors.ex [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+
+        * `:distance` - weight points by the inverse of their distance. in this case, closer neighbors of
+          a query point will have a greater influence than neighbors which are further away.
+      """
+    ],
+    metric: [
+      type: {:custom, Scholar.Options, :metric, []},
+      default: {:minkowski, 2},
+      doc: ~S"""
+      Name of the metric. Possible values:
```
Comment (josevalim) [inline lib/scholar/neighbors/k_nearest_neighbors.ex:40]:
```suggestion
      Name of the metric. Possible values:

```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**